### PR TITLE
Adds support for 3DTILES_implicit_tiling

### PR DESCRIPTION
**Pull request by ErixenCruz**
Need to fix the bounding volumes as described in https://github.com/CesiumGS/cdb-to-3dtiles/issues/39. Also need to add the negative levels of detail as external tilesets along with the implicitly defined positive geocell levels and complete variable subtree levels support.

### CODE REVIEW

**Comment by ErixenCruz** [comment]
All that is left to do is fix some hard coding of URIs in subtree, tweak morton index calculation to account for variable subtree level, better comments and documentation and tests, and double check when running full Yemen. Eventually multi-content

File: CDBTo3DTiles/CMakeLists.txt [hunk 1]
```diff
@@ -28,15 +29,39 @@ set(PRIVATE_THIRD_PARTY_INCLUDE_PATHS
     ${nlohmann_json_INCLUDE_DIRS}
     ${GDAL_INCLUDE_DIRS}
     ${tinygltf_INCLUDE_DIRS}
+    ${LIBMORTON_INCLUDE_DIR}
+    ${glm_INCLUDE_DIR}
 )
 
-target_include_directories(CDBTo3DTiles
+get_target_property(VAR CDBTo3DTiles CDBTo3DTiles_THIRD_PARTY_INCLUDE_PRIVATE)
+message("This is from line 38:" ${VAR})
```
Comment (sanjeetsuhag) [inline CDBTo3DTiles/CMakeLists.txt:37]: This debug statement should be removed.

File: CDBTo3DTiles/include/CDBTo3DTiles.h [hunk 2]
```diff
@@ -4,6 +4,8 @@
 #include <memory>
 #include <string>
 #include <vector>
+#include "../src/ConverterImpl.h"
+// #include "ConverterImpl.h"
```
Comment (sanjeetsuhag) [inline CDBTo3DTiles/include/CDBTo3DTiles.h:8]: Remove unnecessary comments.

File: CDBTo3DTiles/include/CDBTo3DTiles.h [hunk 3]
```diff
@@ -34,9 +40,9 @@ class Converter
     void convert();
 
 private:
-    struct Impl;
+    // struct Impl;
```
Comment (sanjeetsuhag) [inline CDBTo3DTiles/include/CDBTo3DTiles.h:43]: Remove unnecessary comments.

File: Core/src/FileUtil.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
+#include "FileUtil.h"
+
+#include <fstream>
+#include <filesystem>
+namespace fs = std::filesystem;
+
+namespace AGI
```
Comment (sanjeetsuhag) [inline Core/src/FileUtil.cpp:7]: Change this namespace to `CDBTo3DTiles`

File: Core/include/FileUtil.h [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include <filesystem>
+namespace fs = std::filesystem;
+
+namespace AGI
```
Comment (sanjeetsuhag) [inline Core/include/FileUtil.h:6]: Change namespace to `CDBTo3DTiles`

File: CDBTo3DTiles/src/TileFormatIO.cpp [hunk 6]
```diff
@@ -381,22 +390,42 @@ void convertTilesetToJson(const CDBTile &tile, float geometricError, nlohmann::j
...
+        implicitJson["extensions"] = nlohmann::json::object();
+        implicitJson["content"] = nlohmann::json::object();
+        implicitJson["content"]["uri"] = geoCell.getLatitudeDirectoryName() + geoCell.getLongitudeDirectoryName() + "_D001_S001_T001_L{level}_U{y}_R{x}.b3dm";
+        
+        nlohmann::json implicitTiling;
+        implicitTiling["maximumLevel"] = maxLevel;
+        implicitTiling["subdivisionScheme"] = "QUADTREE";
+        implicitTiling["subtreeLevels"] = subtreeLevels;
+        implicitTiling["subtrees"] = nlohmann::json::object();
+        // TODO fix .. here
```
Comment (sanjeetsuhag) [inline CDBTo3DTiles/src/TileFormatIO.cpp:410]: Is this fix in? Can this comment be removed?

File: CDBTo3DTiles/src/TileFormatIO.cpp [hunk 7]
```diff
@@ -74,9 +75,17 @@ void writeToTilesetJson(const CDBTileset &tileset, bool replace, std::ofstream &
         tilesetJson["root"]["refine"] = "ADD";
     }
 
+    if(threeDTilesNext)
+    {
+      tilesetJson["extensionsUsed"] = nlohmann::json::array();
```
Comment (sanjeetsuhag) [inline CDBTo3DTiles/src/TileFormatIO.cpp:80]: You can add the value in the constructor:

```suggestion
      tilesetJson["extensionsUsed"] = nlohmann::json::array({ "3DTILES_implicit_tiling" });
```

File: CDBTo3DTiles/src/CDBTo3DTiles.cpp [hunk 8]
```diff
@@ -739,117 +111,307 @@ void Converter::convert()
...
-            m_impl->addVectorToTilesetCollection(railRoadNetwork,
-                                                 railRoadNetworkDir,
-                                                 m_impl->railRoadNetworkTilesets);
-        });
-        m_impl->flushTilesetCollection(geoCell, m_impl->railRoadNetworkTilesets);
+          });
+          m_impl->flushTilesetCollection(geoCell, m_impl->elevationTilesets);
+          std::unordered_map<CDBTile, Texture>().swap(m_impl->processedParentImagery);
+
+          // TODO check this logic for multi subtree (buffers)
```
Comment (sanjeetsuhag) [inline CDBTo3DTiles/src/CDBTo3DTiles.cpp:183]: Is this fix in?

File: CDBTo3DTiles/src/CDBTile.h [hunk 9]
```diff
@@ -74,6 +78,8 @@ class CDBTile
 
     std::string getLevelInFilename() const noexcept;
 
+    std::string getLevelWithNoZeroPaddingInFilename() const noexcept;
```
Comment (sanjeetsuhag) [inline CDBTo3DTiles/src/CDBTile.h:81]: Add unit tests for the new functions in this file.

**Review by sanjeetsuhag — CHANGES_REQUESTED** [review]
I did an initial pass, just looking at the diffs.

File: CDBTo3DTiles/src/CDBTo3DTiles.cpp [hunk 10]
```diff
@@ -739,117 +111,304 @@ void Converter::convert()
...
-            auto tilesetAggregateRegion = aggregateTilesetsRegion.find(combinedTilesetName);
-            if (tilesetAggregateRegion == aggregateTilesetsRegion.end()) {
-                aggregateTilesetsRegion.insert({combinedTilesetName, geoCellRegion});
-            } else {
-                tilesetAggregateRegion->second = tilesetAggregateRegion->second.computeUnion(geoCellRegion);
-            }
+        // combine all the default tileset in each geocell into a global one
+        for (auto const& [tilesetName, tileset] : combinedTilesets) {
+            // std::ofstream fs(m_impl->outputPath / (tileset.first + ".json"));
+            // combineTilesetJson(tileset.second, combinedTilesetsRegions[tileset.first], fs);
```
Comment (sanjeetsuhag) [inline CDBTo3DTiles/src/CDBTo3DTiles.cpp:289]: Can these be removed?

File: Core/include/FileUtil.h [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    namespace Utilities
+    {
+        void writeBinaryFile(const fs::path& filename, const char* binary, uint64_t byteLength);
+        void writeBinaryFile(std::ostream& outputStream, const char* binary, uint64_t byteLength);
+        void writeStringToFile(const fs::path& filename, const std::string& string);
+        void readBinaryFile(const fs::path& filename, uint8_t* binary, uint64_t byteLength);
+        std::string readFileToString(const fs::path& filename);
+        size_t getDirectorySize(const fs::path& path);
+    } // namespace Utilities
+} // namespace AGI
```
Comment (sanjeetsuhag) [inline Core/include/FileUtil.h:17]: Replace AGI namespace

File: Core/src/FileUtil.cpp [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+                    {
+                        totalSize += fs::file_size(entry);
+                    }
+                }
+            }
+
+            return totalSize;
+        }
+    } // namespace Utilities
+} // namespace AGI
```
Comment (sanjeetsuhag) [inline Core/src/FileUtil.cpp:75]: Replace AGI namespace

**Review by sanjeetsuhag — CHANGES_REQUESTED** [review]
I'll do a more thorough review of the test cases after the `SEGFAULT` is fixed.

File: CDBTo3DTiles/src/ConverterImpl.cpp [hunk 13]
```diff
@@ -0,0 +1,634 @@
...
+  }
+  if(subtreeLevels < 1) 
+  {
+    throw std::invalid_argument("Subtree level must be positive.");
+  }
+  const auto &cdbTile = elevation.getTile();
+  int level = cdbTile.getLevel();
+  int levelWithinSubtree = level - subtreeRootLevel;
+
+  // TODO the rref and uref need to be with respect to subtree, not larger tree
```
Comment (sanjeetsuhag) [inline CDBTo3DTiles/src/ConverterImpl.cpp:86]: Is this still `TODO`?

File: CDBTo3DTiles/src/ConverterImpl.cpp [hunk 14]
```diff
@@ -0,0 +1,634 @@
...
+    }
+
+    if (cdbTile.getLevel() < 0) {
+        fillMissingNegativeLODElevation(elevation, cdb, tilesetDirectory, tileset);
+    } else {
+        fillMissingPositiveLODElevation(elevation, imagery, cdb, tilesetDirectory, tileset);
+    }
+}
+
+// TODO remove this function?
```
Comment (sanjeetsuhag) [inline CDBTo3DTiles/src/ConverterImpl.cpp:233]: Is this still `TODO`?

File: Tests/CombineTilesetsTest.cpp [hunk 15]
```diff
@@ -407,3 +415,174 @@ TEST_CASE("Test combine multiple sets of tilesets", "[CombineTilesets]")
...
+  std::optional<CDBElevation> elevation = CDBElevation::createFromFile(elevationTilePath);
+  SECTION("Test converter errors out of 3D Tiles Next conversion with uninitialized availabilty buffer.")
+  {
+    uint8_t* nullPointer = NULL;
+    uint64_t dummyInt;
+    REQUIRE_THROWS_AS(m_impl->addElevationAvailability(*elevation, cdb, nullPointer, nullPointer, &dummyInt, &dummyInt, 0, 0, 0), std::invalid_argument);
+  }
+
+  
+  // TODO write function for creating buffer given subtree level
```
Comment (sanjeetsuhag) [inline Tests/CombineTilesetsTest.cpp:436]: Is this still `TODO`?

**Review by sanjeetsuhag — APPROVED** [review]

**Pull request merged**